### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by reegeek** [oxyplot/oxyplot#1658]
Hi,

First oxyplot is a very complete library, thank for your work.

The latest version is almost a year old and a lot of work has been done since then. Why ?
Do you have in mind a releases roadmap ?

Regards.

**Comment by VisualMelon** [comment]
@objorke would have to say, but he has little time available these days. There are milestones for v3 (https://github.com/oxyplot/oxyplot/milestone/3), for which there is a lot of design work to be done still.

See #1615 for notes on how to access CI builds. They are perfectly stable, but APIs are liable to change suddenly if you do use them (keep an eye on the changelog).

**Comment by Jonarw** [comment]
I think this is a very good question which is also hard to answer. Finding the right balance between releasing frequently and avoiding excessive breaking changes in the official releases is a non-trivial task after all.

We do have the milestones for v3 as @VisualMelon mentioned, but as he also hinted there are some issues on there which I don't see being solved in the near future.
However if we feel that we made enough progress since the last release to warrant a new version, personally I would also have no problem with pushing stuff to a future milestone. This is certainly preferable to postponing a potential release indefinitely.

Are we at that point now where there are enough improvements since v2 to warrant a new release? There are some [open](https://github.com/oxyplot/oxyplot/pull/1609) [topics](https://github.com/oxyplot/oxyplot/pull/1556) which personally I think should be part of a potential release, but after that maybe...? But that is where @objorke's input would be welcome :).

**Comment by objorke** [comment]
Also see the wiki roadmap page https://github.com/oxyplot/oxyplot/wiki/Project-Roadmap
 
I think v3 included features I think should be prioritised when making a release with breaking changes. The list should be reviewed again, I think @VisualMelon and @Jonarw could handle that!

I have no objections if you would like to change to a more regular schedule of releases, particularly on non-breaking changes. 
I tried to document a release process at https://github.com/oxyplot/oxyplot/wiki/Release-process
The manual steps takes some time... Please follow/update that document. I can give access to push to NuGet.

I also like release strategies like the one for ember: https://emberjs.com/releases/

And, should there be a "manifesto" (I don´t know if that´s the right word) that defines the main characteristics and overarching goals for the library?

**Comment by VisualMelon** [comment]
Linking #1664: as a start, I think a 2.1 with Skia would be a good idea before I break everything with #1556. 

I'd be keen to discuss this further; I don't know if a manifesto is necessary, but additional (user facing) documentation concerning the goals would be good (I think @Jonarw and I pretty much get the idea, but something definitive to point when people make feature requests would be useful).

**Comment by VisualMelon** [comment]
@Jonarw would you be up for agreeing a 2.1 release with a process something like:
 1. Release a 2.1-preview1 directly from a release2.1 branch
 2. Hope some people use it (I will for my internal stuff, though the obsolescence warnings will be irritating) 
 3. Wait for some bug reports etc. so that it's vaguely stable, and only merge fixes into the 2.1 branch
 4. After a while (few weeks, or more if there are lots of bugs), release 2.1 proper

If you think this is a sensible idea, we can open a new issue to decide what we want to get done before we do step (1), and then take it from there.

The big things in the release (from my perspective) would be:
 - Skia PlotView and Exporters
 - ImageSharp Exporters
 - Consolidated Clipping and improved Rendering APIs
 - Axis Margins (though it might just be me that really wants those)

It would also provide a good step for bringing OxyplotAvalonia and OxyplotGtk up to date.

**Comment by VisualMelon** [comment]
Having a deadline of sorts might also encourage me to get on with writing some more documentation...

**Comment by Jonarw** [comment]
I think your plan sounds quite sensible!

**Comment by oldhamir** [comment]
Is XAML support being removed/deprecated from OxyPlot?

I use XAML a lot as I use MvvmLight

**Comment by VisualMelon** [comment]
@oldhamir They are moved to https://github.com/oxyplot/oxyplot-contrib, though I see there hasn't been NuGet release of that for some years. We should consider 'rebooting' that for the 2.1 release to provide compatibility.

**Comment by VisualMelon** [comment]
Note: removing those demands a significant update to the WPF documentation.

**Comment by oldhamir** [comment]
So is the intention to continue XAML support going forward? If it's going to die out in v4.0 then I will either have to re-write what I've currently done or find another lib...

I had pulled the oxyplot-contrib repo and noted that for instance the Plot class cannot find the PlotBase class anywhere hence my original question. I would be willing to provide assistance in keeping OxyPlot going, although I've never contributed to an Open Source project before. I've read the Contributing.md but at least initially could do with some pointers as for instance I don't know what the intention is by removing that PlotBase class. I suspect that it's been renamed to PlotViewBase but it's not a straight shoe-in for PlotBase, I already tried that...

**Comment by oldhamir** [comment]
When I say XAML support I mean more that just declaring a PlotView in the XAML file. E.g. this is an example of what I can currently do with OxyPlot v2:

        <oxyPlot:Plot x:Name="temporalHistoryGraph" Title="{Binding TheGraphDetails.GraphTitle}" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" IsLegendVisible="False" 
                      SizeChanged="Graph_SizeChanged" IsMouseWheelEnabled="False" IsManipulationEnabled="False" BorderThickness="0,0,16,0" Cursor="Cross" 
                      PreviewMouseMove="Graph_PreviewMouseMove" MouseRightButtonUp="Graph_MouseRightButtonUp" MouseLeftButtonUp="Graph_MouseLeftButtonUp" MouseDoubleClick="Graph_MouseDoubleClick" 
                      MouseLeftButtonDown="Graph_MouseLeftButtonDown">
            <oxyPlot:Plot.Axes>
                <oxyPlot:DateTimeAxis x:Name="xAxis" Position="Bottom" Minimum="{Binding TheGraphDetails.MinXValue}" Maximum="{Binding TheGraphDetails.MaxXValue}" TitleFontWeight="Bold" Title="Log Timestamp" IsPanEnabled="False" IsZoomEnabled="False" />
                <oxyPlot:LinearAxis x:Name="yAxis" Position="Left" Minimum="{Binding TheGraphDetails.MinYValue}" Maximum="{Binding TheGraphDetails.MaxYValue}" TitleFontWeight="Bold" Title="RPM" IsZoomEnabled="False" IsPanEnabled="False" />
            </oxyPlot:Plot.Axes>
            <oxyPlot:Plot.Series>
                <oxyPlot:ScatterSeries x:Name="alarmWarnSeries" ItemsSource="{Binding TheGraphDetails.HighlightedEntries1}" DataFieldX="X" DataFieldY="Y" MarkerSize="3" MarkerType="Circle" MarkerFill="OrangeRed" />
                <oxyPlot:ScatterSeries x:Name="allEntrySeries" ItemsSource="{Binding TheGraphDetails.AllEntries1}" DataFieldX="X" DataFieldY="Y" MarkerSize="2" MarkerType="Diamond" MarkerFill="Black" />
            </oxyPlot:Plot.Series>
        </oxyPlot:Plot>

**Comment by VisualMelon** [comment]
@oldhamir Yep, I'm just taking a look myself, and the contrib stuff isn't in a fit state at the moment. I'll try to get it at least building today.

The problem with the XAML stuff is that only a small number of people understand it and it's time-consuming to test. Consequently, it's too high a bar to expect contributors to the core library to keep it up to date. I'm not sure there is a plan at all in the long run: we will certainly accept PRs that resolve issues with it, and if there is interest I'll consider taking the time to patch it (normally I put my efforts into things I use, because it's hard to make design decisions about something you've never used). We need to sort out our stance on this, because it applies to Oxyplot-Avalonia as well.

If you do want to contribute, feel free to open issues explaining problems, and either (hopefully) someone will fix them, or I'll be happy to assist with anything (git, design, w/e) when I have time.

<hr />

@Jonarw we don't have _any_ information on usage beyond the nuget download counts, and they may not be very indicative (e.g. I only started uses NuGet builds with 2.0, before that I was exclusively using custom builds). Would it be worth arranging some sort of low-effort anonymous survey to gauge demand, and help focus 'surplus' (i.e. stuff people do that isn't for their own purposes) effort?

**Comment by Jonarw** [comment]
(To avoid confusion: I saw and commented on https://github.com/oxyplot/oxyplot-contrib/pull/3 before I saw this, so here I will address the actual points some more).

In my mind, the XAML bindings for WPF are 'nice-to-have' in the sense that they provide an alternative way to specify plots on WPF, which might be convenient for some applications.

But in the end, in my opinion they give no added value over just specifying the plot 'normally' in C# code and are also much more limited in their usage. They also could lead people into a dead-end situation, where after starting out with XAML they realize they need to support WinForms as well, or need the PDF exporter from `OxyPlot.SkiaSharp` (or want to use a newer version of OxyPlot which doesn't support the XAML stuff anymore - which is kind of @oldhamir's situation).

Also they are kind of inconvenient to maintain and test.
So in my opinion we shouldn't lose any sleep over supporting these further. Of course they will continue to be available with oxyplot 2.0, and maybe also for future versions if @VisualMelon (or others) are motivated enough to keep them going.

**Comment by VisualMelon** [comment]
@Jonarw Yeah, I think I second all that. I'll do my best to get contrib in working shape for 2.1 so that if there _is_ demand it's in a state such that people like @oldhamir can take up the torch if they deem it worthwhile without having to fight with a lump of non-compiling code.
